File: db/examples_cxx/AppinitExample.cpp
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997, 1998
 *	Sleepycat Software.  All rights reserved.
 *
 *	@(#)AppinitExample.cpp	10.6 (Sleepycat) 4/10/98
 */

#include "config.h"

#ifndef NO_SYSTEM_INCLUDES
#include <sys/types.h>

#include <errno.h>
#include <iostream.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#endif

#include <db_cxx.h>

#define	DATABASE_HOME	"/home/database"

DB_ENV *db_setup(FILE *);

char *progname = "AppinitExample";			/* Program name. */

/*
 * An example of a program using DbEnv to configure its DB
 * environment.
 */
DbEnv *
init()
{
    DbEnv *dbenv = new DbEnv();
    char *config[2];

    // Output errors to the application's log.
    //
    dbenv->set_error_stream(&cerr);
    dbenv->set_errpfx(progname);

    //
    // All of the shared database files live in /home/database,
    // but data files live in /database.
    //
    config[0] = "DB_DATA_DIR /database/files";
    config[1] = NULL;

    //
    // We want to specify the shared memory buffer pool cachesize,
    // but everything else is the default.
    //
    dbenv->set_mp_size(64 * 1024);

    //
    // We have multiple processes reading/writing these files, so
    // we need concurrency control and a shared buffer pool, but
    // not logging or transactions.
    //
    if ((errno = dbenv->appinit(DATABASE_HOME,
	    config, DB_CREATE | DB_INIT_LOCK | DB_INIT_MPOOL)) != 0) {
        cerr << progname << ": db_appinit: " << strerror(errno);
        exit (1);
    }
    return dbenv;
}

int
main(int, char **)
{
    //
    // Note: it may be easiest to put all DB operations in a try block, as
    // seen here.  Alternatively, you can change the ErrorModel in the DbEnv
    // so that exceptions are never thrown and check error returns from all
    // methods.
    //
    try
    {
        DbEnv *dbenv = init();

        // ... your application here ...

        delete dbenv;
        return 0;
    }
    catch (DbException &dbe)
    {
        cerr << "AccessExample: " << dbe.what() << "\n";
        return 1;
    }
}

File: db/java/src/com/sleepycat/examples/AppinitExample.java
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997, 1998
 *	Sleepycat Software.  All rights reserved.
 *
 *	@(#)AppinitExample.java	10.3 (Sleepycat) 4/10/98
 */

package com.sleepycat.examples;

import com.sleepycat.db.*;

/*
 * An example of a program using DbEnv to configure its DB
 * environment.
 *
 * For comparison purposes, this example uses a similar structure
 * as examples/ex_appinit.c and examples_cxx/AppinitExample.cpp.
 */
class AppinitExample
{
    private static final String progname = "AppinitExample";
    private static final String DATABASE_HOME = "/home/database";

    private static DbEnv init()
         throws DbException
    {
        DbEnv dbenv = new DbEnv();
        String config[] = new String[1];

        // Output errors to the application's log.
        //
        dbenv.set_error_stream(System.err);
        dbenv.set_errpfx(progname);

        //
        // All of the shared database files live in /home/database,
        // but data files live in /database.
        //
        // In C/C++ we need to allocate two elements in the array
        // and set config[1] to NULL.  This is not necessary in Java.
        //
        config[0] = "DB_DATA_DIR /database/files";

        //
        // We want to specify the shared memory buffer pool cachesize,
        // but everything else is the default.
        //
        dbenv.set_mp_size(64 * 1024);

        //
        // We have multiple processes reading/writing these files, so
        // we need concurrency control and a shared buffer pool, but
        // not logging or transactions.
        //
        // appinit() will throw on error.
        //
        dbenv.appinit(DATABASE_HOME, config,
                      Db.DB_CREATE | Db.DB_INIT_LOCK | Db.DB_INIT_MPOOL);
        return dbenv;
    }

    static public void main(String[] args)
    {
	System.out.println("beginning appinit");
        try {
            DbEnv dbenv = init();

            // ... your application here ...

            dbenv.appexit();
        }
        catch (DbException dbe) {
            System.err.println(progname + ": db_appinit: " + dbe.toString());
            System.exit (1);
        }
	System.out.println("completed appinit");
    }
}
